File: ios/MediaClipboard.m
#import "MediaClipboard.h"
#import <React/RCTLog.h>
#import <UIKit/UIKit.h>
#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>
#import <MobileCoreServices/MobileCoreServices.h>

@implementation MediaClipboard

RCT_EXPORT_MODULE()

// Indicate that this module wants to be initialized on the main queue
+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

#pragma mark - React Native Methods

RCT_EXPORT_METHOD(copyText:(NSString *)text
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    UIPasteboard.generalPasteboard.string = text;
    resolve(nil);
  });
}

RCT_EXPORT_METHOD(copyImage:(NSString *)imagePath
                  options:(NSDictionary *)options
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    @try {
      UIImage *image = [self loadImageFromPath:imagePath];
      if (image) {
        dispatch_async(dispatch_get_main_queue(), ^{
          UIPasteboard.generalPasteboard.image = image;
          resolve(nil);
        });
      } else {
        NSString *resolvedPath = [self normalizeFilePath:imagePath];
        NSString *errorMessage;
        
        // Check if this is a relative path issue
        if (!resolvedPath && ([imagePath hasPrefix:@"./"] || [imagePath hasPrefix:@"../"] || [imagePath containsString:@"/../"] || [imagePath containsString:@"/./"]) ) {
          errorMessage = [NSString stringWithFormat:@"Cannot copy from relative path '%@'. Relative paths like './' and '../' are not supported. Please use one of these alternatives:\n1. Convert to base64: data:image/jpeg;base64,<your_base64_data>\n2. Use HTTP/HTTPS URL: https://example.com/image.jpg\n3. Use absolute file path: /absolute/path/to/image.jpg\n4. Use file:// URL: file:///absolute/path/to/image.jpg", imagePath];
        } else {
          errorMessage = resolvedPath ? 
            [NSString stringWithFormat:@"Cannot copy image from resolved path: %@", resolvedPath] :
            [NSString stringWithFormat:@"Cannot resolve or copy image from path: %@", imagePath];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
          reject(@"COPY_IMAGE_ERROR", errorMessage, nil);
        });
      }
    } @catch (NSException *exception) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"COPY_IMAGE_ERROR", exception.reason, nil);
      });
    }
  });
}

RCT_EXPORT_METHOD(copyVideo:(NSString *)videoPath
                  options:(NSDictionary *)options
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    @try {
      NSData *videoData = [self loadFileDataFromPath:videoPath];
      if (videoData) {
        dispatch_async(dispatch_get_main_queue(), ^{
          if (@available(iOS 14.0, *)) {
            UIPasteboard.generalPasteboard.items = @[@{UTTypeMovie.identifier: videoData}];
          } else {
            [UIPasteboard.generalPasteboard setData:videoData forPasteboardType:(__bridge NSString *)kUTTypeMovie];
          }
          resolve(nil);
        });
      } else {
        NSString *errorMessage;
        
        // Check if this is a relative path issue
        if ([videoPath hasPrefix:@"./"] || [videoPath hasPrefix:@"../"] || [videoPath containsString:@"/../"] || [videoPath containsString:@"/./"]) {
          errorMessage = [NSString stringWithFormat:@"Cannot copy from relative path '%@'. Relative paths like './' and '../' are not supported. Please use one of these alternatives:\n1. Use HTTP/HTTPS URL: https://example.com/video.mp4\n2. Use absolute file path: /absolute/path/to/video.mp4\n3. Use file:// URL: file:///absolute/path/to/video.mp4", videoPath];
        } else {
          errorMessage = [NSString stringWithFormat:@"Failed to load video from path: %@", videoPath];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
          reject(@"COPY_VIDEO_ERROR", errorMessage, nil);
        });
      }
    } @catch (NSException *exception) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"COPY_VIDEO_ERROR", exception.reason, nil);
      });
    }
  });
}

RCT_EXPORT_METHOD(copyPDF:(NSString *)pdfPath
                 options:(NSDictionary *)options
                withResolver:(RCTPromiseResolveBlock)resolve
                withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    @try {
      NSData *pdfData = [self loadFileDataFromPath:pdfPath];
      if (pdfData) {
        dispatch_async(dispatch_get_main_queue(), ^{
          if (@available(iOS 14.0, *)) {
            UIPasteboard.generalPasteboard.items = @[@{UTTypePDF.identifier: pdfData}];
          } else {
            [UIPasteboard.generalPasteboard setData:pdfData forPasteboardType:(__bridge NSString *)kUTTypePDF];
          }
          resolve(nil);
        });
      } else {
        NSString *errorMessage;
        
        // Check if this is a relative path issue
        if ([pdfPath hasPrefix:@"./"] || [pdfPath hasPrefix:@"../"] || [pdfPath containsString:@"/../"] || [pdfPath containsString:@"/./"]) {
          errorMessage = [NSString stringWithFormat:@"Cannot copy from relative path '%@'. Relative paths like './' and '../' are not supported. Please use one of these alternatives:\n1. Use HTTP/HTTPS URL: https://example.com/document.pdf\n2. Use absolute file path: /absolute/path/to/document.pdf\n3. Use file:// URL: file:///absolute/path/to/document.pdf", pdfPath];
        } else {
          errorMessage = [NSString stringWithFormat:@"Failed to load PDF from path: %@", pdfPath];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
          reject(@"COPY_PDF_ERROR", errorMessage, nil);
        });
      }
    } @catch (NSException *exception) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"COPY_PDF_ERROR", exception.reason, nil);
      });
    }
  });
}

RCT_EXPORT_METHOD(copyAudio:(NSString *)audioPath
                  options:(NSDictionary *)options
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    @try {
      NSData *audioData = [self loadFileDataFromPath:audioPath];
      if (audioData) {
        NSString *audioType = [self getAudioUTTypeForPath:audioPath];
        dispatch_async(dispatch_get_main_queue(), ^{
          if (@available(iOS 14.0, *)) {
            UIPasteboard.generalPasteboard.items = @[@{audioType: audioData}];
          } else {
            [UIPasteboard.generalPasteboard setData:audioData forPasteboardType:audioType];
          }
          resolve(nil);
        });
      } else {
        NSString *errorMessage;
        
        // Check if this is a relative path issue
        if ([audioPath hasPrefix:@"./"] || [audioPath hasPrefix:@"../"] || [audioPath containsString:@"/../"] || [audioPath containsString:@"/./"]) {
          errorMessage = [NSString stringWithFormat:@"Cannot copy from relative path '%@'. Relative paths like './' and '../' are not supported. Please use one of these alternatives:\n1. Use HTTP/HTTPS URL: https://example.com/audio.mp3\n2. Use absolute file path: /absolute/path/to/audio.mp3\n3. Use file:// URL: file:///absolute/path/to/audio.mp3", audioPath];
        } else {
          errorMessage = [NSString stringWithFormat:@"Failed to load audio from path: %@", audioPath];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
          reject(@"COPY_AUDIO_ERROR", errorMessage, nil);
        });
      }
    } @catch (NSException *exception) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"COPY_AUDIO_ERROR", exception.reason, nil);
      });
    }
  });
}

RCT_EXPORT_METHOD(copyFile:(NSString *)filePath
                  mimeType:(NSString *)mimeType
                  options:(NSDictionary *)options
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    @try {
      NSData *fileData = [self loadFileDataFromPath:filePath];
      if (fileData) {
        NSString *utType = [self getUTTypeForMimeType:mimeType];
        dispatch_async(dispatch_get_main_queue(), ^{
          if (@available(iOS 14.0, *)) {
            UIPasteboard.generalPasteboard.items = @[@{utType: fileData}];
          } else {
            [UIPasteboard.generalPasteboard setData:fileData forPasteboardType:utType];
          }
          resolve(nil);
        });
      } else {
        NSString *errorMessage;
        
        // Check if this is a relative path issue
        if ([filePath hasPrefix:@"./"] || [filePath hasPrefix:@"../"] || [filePath containsString:@"/../"] || [filePath containsString:@"/./"]) {
          errorMessage = [NSString stringWithFormat:@"Cannot copy from relative path '%@'. Relative paths like './' and '../' are not supported. Please use one of these alternatives:\n1. Convert to base64: data:%@;base64,<your_base64_data>\n2. Use HTTP/HTTPS URL: https://example.com/file\n3. Use absolute file path: /absolute/path/to/file\n4. Use file:// URL: file:///absolute/path/to/file", filePath, mimeType];
        } else {
          errorMessage = [NSString stringWithFormat:@"Failed to load file from path: %@", filePath];
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
          reject(@"COPY_FILE_ERROR", errorMessage, nil);
        });
      }
    } @catch (NSException *exception) {
      dispatch_async(dispatch_get_main_queue(), ^{
        reject(@"COPY_FILE_ERROR", exception.reason, nil);
      });
    }
  });
}

RCT_EXPORT_METHOD(copyLargeFile:(NSString *)filePath
                       mimeType:(NSString *)mimeType
                       options:(NSDictionary *)options
                      withResolver:(RCTPromiseResolveBlock)resolve
                      withRejecter:(RCTPromiseRejectBlock)reject)
{
  // For large files, we'll use the same implementation as copyFile for simplicity
  [self copyFile:filePath mimeType:mimeType options:options withResolver:resolve withRejecter:reject];
}

RCT_EXPORT_METHOD(hasContent:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    BOOL hasContent = UIPasteboard.generalPasteboard.items.count > 0;
    resolve(@(hasContent));
  });
}

RCT_EXPORT_METHOD(getContent:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    UIPasteboard *pasteboard = UIPasteboard.generalPasteboard;
    NSMutableDictionary *content = [NSMutableDictionary dictionary];
    
    if (pasteboard.string) {
      content[@"type"] = @"text";
      content[@"data"] = pasteboard.string;
    } else if (pasteboard.image) {
      content[@"type"] = @"image";
      NSData *imageData = UIImagePNGRepresentation(pasteboard.image);
      if (imageData) {
        content[@"size"] = @(imageData.length);
      }
    } else if (pasteboard.hasImages) {
      content[@"type"] = @"image";
    } else if (pasteboard.hasURLs) {
      content[@"type"] = @"text";
      content[@"data"] = pasteboard.URL.absoluteString;
    } else if (pasteboard.items.count > 0) {
      content[@"type"] = @"file";
      content[@"size"] = @(pasteboard.items.count);
    } else {
      content[@"type"] = @"unknown";
    }
    
    resolve(content);
  });
}

RCT_EXPORT_METHOD(clear:(RCTPromiseResolveBlock)resolve
          withRejecter:(RCTPromiseRejectBlock)reject)
{
  dispatch_async(dispatch_get_main_queue(), ^{
    UIPasteboard.generalPasteboard.items = @[];
    resolve(nil);
  });
}

#pragma mark - Helper Methods

- (UIImage *)loadImageFromPath:(NSString *)path {
  // Handle base64 data URLs
  if ([path hasPrefix:@"data:image/"]) {
    NSRange range = [path rangeOfString:@","];
    if (range.location != NSNotFound) {
      NSString *base64String = [path substringFromIndex:range.location + 1];
      NSData *data = [[NSData alloc] initWithBase64EncodedString:base64String options:NSDataBase64DecodingIgnoreUnknownCharacters];
      return [UIImage imageWithData:data];
    }
    return nil;
  }
  
  // Handle HTTP/HTTPS URLs
  if ([path hasPrefix:@"http://"] || [path hasPrefix:@"https://"]) {
    return [self loadImageFromURL:path];
  }
  
  // Normalize file path and handle different URI schemes
  NSString *normalizedPath = [self normalizeFilePath:path];
  if (!normalizedPath) {
    NSLog(@"[MediaClipboard] Failed to resolve path: %@ -> null", path);
    return nil;
  }
  
  UIImage *image = [UIImage imageWithContentsOfFile:normalizedPath];
  if (!image) {
    NSLog(@"[MediaClipboard] Failed to load image from resolved path: %@", normalizedPath);
  }
  
  return image;
}

- (NSData *)loadFileDataFromPath:(NSString *)path {
  // Handle HTTP/HTTPS URLs
  if ([path hasPrefix:@"http://"] || [path hasPrefix:@"https://"]) {
    return [self loadDataFromURL:path];
  }
  
  // Normalize file path and handle different URI schemes
  NSString *normalizedPath = [self normalizeFilePath:path];
  if (!normalizedPath) {
    NSLog(@"[MediaClipboard] Failed to resolve path: %@ -> null", path);
    return nil;
  }
  
  if (![[NSFileManager defaultManager] fileExistsAtPath:normalizedPath]) {
    NSLog(@"[MediaClipboard] File does not exist at resolved path: %@", normalizedPath);
    return nil;
  }
  
  NSData *data = [NSData dataWithContentsOfFile:normalizedPath];
  if (!data) {
    NSLog(@"[MediaClipboard] Failed to load data from resolved path: %@", normalizedPath);
  }
  
  return data;
}

- (NSString *)getUTTypeForMimeType:(NSString *)mimeType {
  if (@available(iOS 14.0, *)) {
    UTType *utType = [UTType typeWithMIMEType:mimeType];
    return utType ? utType.identifier : UTTypeData.identifier;
  } else {
    // Fallback for older iOS versions
    if ([mimeType isEqualToString:@"image/jpeg"] || [mimeType isEqualToString:@"image/jpg"]) {
      return (__bridge NSString *)kUTTypeJPEG;
    } else if ([mimeType isEqualToString:@"image/png"]) {
      return (__bridge NSString *)kUTTypePNG;
    } else if ([mimeType isEqualToString:@"video/mp4"]) {
      return (__bridge NSString *)kUTTypeMPEG4;
    } else if ([mimeType isEqualToString:@"application/pdf"]) {
      return (__bridge NSString *)kUTTypePDF;
    } else if ([mimeType isEqualToString:@"audio/mpeg"] || [mimeType isEqualToString:@"audio/mp3"]) {
      return (__bridge NSString *)kUTTypeMP3;
    } else if ([mimeType isEqualToString:@"audio/wav"]) {
      return (__bridge NSString *)kUTTypeWaveformAudio;
    }
    return (__bridge NSString *)kUTTypeData;
  }
}

- (NSString *)getAudioUTTypeForPath:(NSString *)audioPath {
  NSString *pathExtension = [audioPath.pathExtension lowercaseString];
  
  if (@available(iOS 14.0, *)) {
    if ([pathExtension isEqualToString:@"mp3"]) {
      return UTTypeMP3.identifier;
    } else if ([pathExtension isEqualToString:@"wav"]) {
      return UTTypeWAV.identifier;
    } else if ([pathExtension isEqualToString:@"m4a"]) {
      return UTTypeMPEG4Audio.identifier;
    } else if ([pathExtension isEqualToString:@"aac"]) {
      return [UTType typeWithIdentifier:@"public.aac"].identifier;
    }
    return UTTypeAudio.identifier;
  } else {
    if ([pathExtension isEqualToString:@"mp3"]) {
      return (__bridge NSString *)kUTTypeMP3;
    } else if ([pathExtension isEqualToString:@"wav"]) {
      return (__bridge NSString *)kUTTypeWaveformAudio;
    } else if ([pathExtension isEqualToString:@"m4a"]) {
      return (__bridge NSString *)kUTTypeMPEG4Audio;
    }
    return (__bridge NSString *)kUTTypeAudio;
  }
}

- (NSString *)normalizeFilePath:(NSString *)path {
  if (!path || path.length == 0) {
    return nil;
  }
  
  // Handle file:// URLs
  if ([path hasPrefix:@"file://"]) {
    NSURL *url = [NSURL URLWithString:path];
    return url.path;
  }
  
  // Handle content:// URLs (Android-style, but sometimes passed to iOS)
  if ([path hasPrefix:@"content://"]) {
    // Content URLs are not directly supported on iOS, but we can try to extract the path
    // This is mainly for compatibility if Android paths are mistakenly passed to iOS
    NSLog(@"Warning: content:// URLs are not supported on iOS: %@", path);
    return nil;
  }
  
  // Handle React Native asset paths (relative paths that might be bundle resources)
  if (![path hasPrefix:@"/"] && ![path hasPrefix:@"file://"] && ![path hasPrefix:@"data:"]) {
    NSLog(@"[MediaClipboard] Processing relative path: %@", path);
    
    // Check for relative paths with dots - these are ambiguous and should be rejected
    if ([path hasPrefix:@"./"] || [path hasPrefix:@"../"] || [path containsString:@"/../"] || [path containsString:@"/./"]) {
      NSLog(@"[MediaClipboard] ERROR: Relative paths with dots are not supported: %@", path);
      return nil; // This will cause the calling method to show an error
    }
    
    // Try multiple resolution strategies for React Native assets
    NSString *resolvedPath = [self resolveReactNativeAssetPath:path];
    if (resolvedPath) {
      return resolvedPath;
    }
    
    NSLog(@"[MediaClipboard] Asset resolution failed, using Documents directory fallback");
    // Fallback: treat as relative path in Documents directory
    NSString *documentsPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject];
    return [documentsPath stringByAppendingPathComponent:path];
  }
  
  // Handle absolute paths as-is
  return path;
}

- (NSString *)resolveReactNativeAssetPath:(NSString *)relativePath {
  NSLog(@"[MediaClipboard] Resolving React Native asset path: %@", relativePath);
  
  // Use the path as-is since dots have already been checked and rejected
  NSString *cleanPath = relativePath;
  
  // Remove leading / if present
  if ([cleanPath hasPrefix:@"/"]) {
    cleanPath = [cleanPath substringFromIndex:1];
  }
  
  NSLog(@"[MediaClipboard] Processing path: %@", cleanPath);
  
  NSFileManager *fileManager = [NSFileManager defaultManager];
  
  // Get common iOS app directories
  NSBundle *mainBundle = [NSBundle mainBundle];
  NSString *bundlePath = [mainBundle bundlePath];
  NSString *documentsPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject];
  NSString *libraryPath = [NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES) firstObject];
  NSString *cachesPath = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) firstObject];
  NSString *tmpPath = NSTemporaryDirectory();
  
  // Strategy 1: Try common React Native asset locations in order of likelihood
  NSArray *basePaths = @[
    bundlePath,                    // Main app bundle
    documentsPath,                 // Documents directory
    [libraryPath stringByAppendingPathComponent:@"Application Support"], // Application Support
    cachesPath,                    // Caches directory  
    tmpPath,                       // Temporary directory
    libraryPath                    // Library directory
  ];
  
  NSArray *subPaths = @[
    cleanPath,                                    // Direct path
    [NSString stringWithFormat:@"assets/%@", cleanPath],          // assets/
    [NSString stringWithFormat:@"src/assets/%@", cleanPath],      // src/assets/
    [NSString stringWithFormat:@"app/assets/%@", cleanPath],      // app/assets/
    [NSString stringWithFormat:@"resources/%@", cleanPath],       // resources/
    [NSString stringWithFormat:@"www/%@", cleanPath],            // www/ (for hybrid apps)
    [NSString stringWithFormat:@"public/%@", cleanPath],         // public/
    [NSString stringWithFormat:@"static/%@", cleanPath]          // static/
  ];
  
  for (NSString *basePath in basePaths) {
    for (NSString *subPath in subPaths) {
      NSString *fullPath = [basePath stringByAppendingPathComponent:subPath];
      if ([fileManager fileExistsAtPath:fullPath]) {
        NSLog(@"[MediaClipboard] Found asset at: %@", fullPath);
        return fullPath;
      }
    }
  }
  
  // Strategy 2: Try NSBundle resource lookup (for truly bundled resources)
  NSString *filename = [cleanPath lastPathComponent];
  NSString *directory = [cleanPath stringByDeletingLastPathComponent];
  NSString *name = [filename stringByDeletingPathExtension];
  NSString *extension = [filename pathExtension];
  
  if (name.length > 0) {
    NSString *resourcePath;
    if (directory.length > 0) {
      resourcePath = [mainBundle pathForResource:name ofType:extension inDirectory:directory];
    } else {
      resourcePath = [mainBundle pathForResource:name ofType:extension];
    }
    
    if (resourcePath && [fileManager fileExistsAtPath:resourcePath]) {
      NSLog(@"[MediaClipboard] Found bundle resource at: %@", resourcePath);
      return resourcePath;
    }
  }
  
  NSLog(@"[MediaClipboard] React Native asset not found: %@", relativePath);
  return nil;
}

- (UIImage *)loadImageFromURL:(NSString *)urlString {
  NSLog(@"[MediaClipboard] Loading image from URL: %@", urlString);
  
  NSURL *url = [NSURL URLWithString:urlString];
  if (!url) {
    NSLog(@"[MediaClipboard] Invalid URL: %@", urlString);
    return nil;
  }
  
  // Synchronous download (for simplicity, in production consider async)
  NSError *error;
  NSData *data = [NSData dataWithContentsOfURL:url options:NSDataReadingUncached error:&error];
  
  if (error) {
    NSLog(@"[MediaClipboard] Failed to download from URL %@: %@", urlString, error.localizedDescription);
    return nil;
  }
  
  if (!data) {
    NSLog(@"[MediaClipboard] No data received from URL: %@", urlString);
    return nil;
  }
  
  UIImage *image = [UIImage imageWithData:data];
  if (!image) {
    NSLog(@"[MediaClipboard] Failed to create image from downloaded data from URL: %@", urlString);
  } else {
    NSLog(@"[MediaClipboard] Successfully loaded image from URL: %@", urlString);
  }
  
  return image;
}

- (NSData *)loadDataFromURL:(NSString *)urlString {
  NSLog(@"[MediaClipboard] Loading data from URL: %@", urlString);
  
  NSURL *url = [NSURL URLWithString:urlString];
  if (!url) {
    NSLog(@"[MediaClipboard] Invalid URL: %@", urlString);
    return nil;
  }
  
  // Synchronous download (for simplicity, in production consider async)
  NSError *error;
  NSData *data = [NSData dataWithContentsOfURL:url options:NSDataReadingUncached error:&error];
  
  if (error) {
    NSLog(@"[MediaClipboard] Failed to download from URL %@: %@", urlString, error.localizedDescription);
    return nil;
  }
  
  if (!data) {
    NSLog(@"[MediaClipboard] No data received from URL: %@", urlString);
    return nil;
  }
  
  NSLog(@"[MediaClipboard] Successfully loaded %lu bytes from URL: %@", (unsigned long)data.length, urlString);
  return data;
}

@end 

File: ios/MediaClipboard.swift
import Foundation
import UIKit
import UniformTypeIdentifiers
import MobileCoreServices

@objc protocol MediaClipboardSwiftDelegate: AnyObject {
    func onProgressUpdate(_ progress: Double)
}

@objc(MediaClipboardSwift)
class MediaClipboardSwift: NSObject {
    
    @objc static let shared = MediaClipboardSwift()
    
    weak var delegate: MediaClipboardSwiftDelegate?
    
    private override init() {
        super.init()
    }
    
    @objc func setDelegate(_ delegate: MediaClipboardSwiftDelegate?) {
        self.delegate = delegate
    }
    
    // MARK: - Public Methods
    
    @objc func copyText(_ text: String, completion: @escaping (Error?) -> Void) {
        DispatchQueue.main.async {
            UIPasteboard.general.string = text
            completion(nil)
        }
    }
    
    @objc func copyImage(_ imagePath: String, options: [String: Any], completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let image = try self.loadImage(from: imagePath)
                DispatchQueue.main.async {
                    UIPasteboard.general.image = image
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    @objc func copyVideo(_ videoPath: String, options: [String: Any], completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let videoData = try self.loadFileData(from: videoPath)
                DispatchQueue.main.async {
                    if #available(iOS 14.0, *) {
                        UIPasteboard.general.items = [[UTType.movie.identifier: videoData]]
                    } else {
                        UIPasteboard.general.setData(videoData, forPasteboardType: kUTTypeMovie as String)
                    }
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    @objc func copyPDF(_ pdfPath: String, options: [String: Any], completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let pdfData = try self.loadFileData(from: pdfPath)
                DispatchQueue.main.async {
                    if #available(iOS 14.0, *) {
                        UIPasteboard.general.items = [[UTType.pdf.identifier: pdfData]]
                    } else {
                        UIPasteboard.general.setData(pdfData, forPasteboardType: kUTTypePDF as String)
                    }
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    @objc func copyAudio(_ audioPath: String, options: [String: Any], completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let audioData = try self.loadFileData(from: audioPath)
                let audioType = self.getAudioUTType(for: audioPath)
                
                DispatchQueue.main.async {
                    if #available(iOS 14.0, *) {
                        UIPasteboard.general.items = [[audioType.identifier: audioData]]
                    } else {
                        UIPasteboard.general.setData(audioData, forPasteboardType: audioType.identifier)
                    }
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    @objc func copyFile(_ filePath: String, mimeType: String, options: [String: Any], completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let fileData = try self.loadFileData(from: filePath)
                let utType = self.getUTType(for: mimeType)
                
                DispatchQueue.main.async {
                    if #available(iOS 14.0, *) {
                        UIPasteboard.general.items = [[utType.identifier: fileData]]
                    } else {
                        UIPasteboard.general.setData(fileData, forPasteboardType: utType.identifier)
                    }
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    @objc func copyLargeFile(_ filePath: String, mimeType: String, options: [String: Any], completion: @escaping (Error?) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let fileData = try self.loadLargeFileData(from: filePath) { [weak self] progress in
                    DispatchQueue.main.async {
                        self?.delegate?.onProgressUpdate(progress)
                    }
                }
                let utType = self.getUTType(for: mimeType)
                
                DispatchQueue.main.async {
                    if #available(iOS 14.0, *) {
                        UIPasteboard.general.items = [[utType.identifier: fileData]]
                    } else {
                        UIPasteboard.general.setData(fileData, forPasteboardType: utType.identifier)
                    }
                    completion(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completion(error)
                }
            }
        }
    }
    
    @objc func hasContent() -> Bool {
        return !UIPasteboard.general.items.isEmpty
    }
    
    @objc(getContentWithCompletion:)
    func getContent(completion: @escaping ([String: Any]?, Error?) -> Void) {
        DispatchQueue.main.async {
            let pasteboard = UIPasteboard.general
            var content: [String: Any] = [:]
            
            if let string = pasteboard.string {
                content["type"] = "text"
                content["data"] = string
            } else if let image = pasteboard.image {
                content["type"] = "image"
                if let imageData = image.pngData() {
                    content["size"] = imageData.count
                }
            } else if pasteboard.hasImages {
                content["type"] = "image"
            } else if pasteboard.hasURLs {
                content["type"] = "text"
                content["data"] = pasteboard.url?.absoluteString
            } else {
                // Check for other data types
                let items = pasteboard.items
                if !items.isEmpty {
                    content["type"] = "file"
                    content["size"] = items.first?.count ?? 0
                }
            }
            
            if content.isEmpty {
                content["type"] = "unknown"
            }
            
            completion(content, nil)
        }
    }
    
    @objc(clearWithCompletion:)
    func clear(completion: @escaping (Error?) -> Void) {
        DispatchQueue.main.async {
            UIPasteboard.general.items = []
            completion(nil)
        }
    }
    
    // MARK: - Private Helper Methods
    
    private func loadImage(from path: String) throws -> UIImage {
        // Handle base64 data URLs
        if path.hasPrefix("data:image/") {
            guard let url = URL(string: path),
                  let data = try? Data(contentsOf: url),
                  let image = UIImage(data: data) else {
                throw MediaClipboardError.invalidImagePath
            }
            return image
        }
        
        // Handle file paths
        let url = URL(fileURLWithPath: path)
        guard let image = UIImage(contentsOfFile: url.path) else {
            throw MediaClipboardError.invalidImagePath
        }
        return image
    }
    
    private func loadFileData(from path: String) throws -> Data {
        let url = URL(fileURLWithPath: path)
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw MediaClipboardError.fileNotFound
        }
        
        do {
            return try Data(contentsOf: url)
        } catch {
            throw MediaClipboardError.fileReadError
        }
    }
    
    private func loadLargeFileData(from path: String, progressCallback: @escaping (Double) -> Void) throws -> Data {
        let url = URL(fileURLWithPath: path)
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw MediaClipboardError.fileNotFound
        }
        
        guard let fileHandle = FileHandle(forReadingAtPath: url.path) else {
            throw MediaClipboardError.fileReadError
        }
        
        defer {
            fileHandle.closeFile()
        }
        
        let fileSize = try FileManager.default.attributesOfItem(atPath: url.path)[.size] as! UInt64
        var data = Data()
        let chunkSize = 1024 * 1024 // 1MB chunks
        var bytesRead: UInt64 = 0
        
        while bytesRead < fileSize {
            let chunk = fileHandle.readData(ofLength: chunkSize)
            if chunk.isEmpty { break }
            
            data.append(chunk)
            bytesRead += UInt64(chunk.count)
            
            let progress = Double(bytesRead) / Double(fileSize)
            progressCallback(progress)
        }
        
        return data
    }
    
    private func getUTType(for mimeType: String) -> UTType {
        if #available(iOS 14.0, *) {
            // iOS 14+ approach - use UTType initializer with mimeType
            if let utType = UTType(mimeType: mimeType) {
                return utType
            }
            return UTType.data
        } else {
            // Fallback for older iOS versions using legacy constants
            switch mimeType {
            case "image/jpeg", "image/jpg":
                return UTType(kUTTypeJPEG as String)!
            case "image/png":
                return UTType(kUTTypePNG as String)!
            case "video/mp4":
                return UTType(kUTTypeMPEG4 as String)!
            case "video/quicktime":
                return UTType(kUTTypeQuickTimeMovie as String)!
            case "application/pdf":
                return UTType(kUTTypePDF as String)!
            case "audio/mpeg", "audio/mp3":
                return UTType(kUTTypeMP3 as String)!
            case "audio/wav":
                return UTType("public.wav")!
            default:
                return UTType(kUTTypeData as String)!
            }
        }
    }
    
    private func getAudioUTType(for audioPath: String) -> UTType {
        let pathExtension = URL(fileURLWithPath: audioPath).pathExtension.lowercased()
        
        if #available(iOS 14.0, *) {
            switch pathExtension {
            case "mp3":
                return UTType.mp3
            case "wav":
                return UTType.wav
            case "m4a":
                return UTType.mpeg4Audio
            case "aac":
                return UTType("public.aac")!
            default:
                return UTType.audio
            }
        } else {
            switch pathExtension {
            case "mp3":
                return UTType(kUTTypeMP3 as String)!
            case "wav":
                return UTType(kUTTypeWaveformAudio as String)!
            case "m4a":
                return UTType(kUTTypeMPEG4Audio as String)!
            default:
                return UTType(kUTTypeAudio as String)!
            }
        }
    }
}

// MARK: - Error Types

enum MediaClipboardError: LocalizedError {
    case invalidImagePath
    case fileNotFound
    case fileReadError
    case unsupportedFormat
    
    var errorDescription: String? {
        switch self {
        case .invalidImagePath:
            return "Invalid image path or data"
        case .fileNotFound:
            return "File not found at specified path"
        case .fileReadError:
            return "Unable to read file data"
        case .unsupportedFormat:
            return "Unsupported file format"
        }
    }
} 